### DISCUSSION TITLE
Idioms For Ad Hoc Polymorphism in `fip` Koka

### DISCUSSION THREAD

**Discussion by d-xo** [koka-lang/koka#345] [General]
Hi!

I've been playing around with some of the examples from the (completely mindblowing :star_struck:) fully in place paper. I wanted to make the `stree` generic on the value type, and quickly ran into trouble. I would want to constrain the allowable types to those that have a total order defined, is there an idiomatic way to do this in koka?

I saw that there was some discussion in https://github.com/koka-lang/koka/issues/16. Has there been any progress since that issue? I tried defining a `total-cmp` effect and using that as described in [this comment](https://github.com/koka-lang/koka/issues/16#issuecomment-304858391) from that thread, but ran into warnings relating to potential allocation in effects other than `exn` and `div`.

It this possible to do in koka at the moment? If not, are there plans to add features supporting ad hoc polymorphism?

**Comment by TimWhiting** [comment]
There are plans to support a form of ad-hoc polymorphism. Daan has started work on a prototype here: https://github.com/koka-lang/koka/tree/dev-implicit

We will see if the prototype turns out to fit Koka well, Daan is in charge, so maybe we will end up with typeclasses eventually. However, implicit parameters seems to work well for ad-hoc polymorphism in Scala, and we believe that our design for implicits works even better for Koka with it's static overloading.

See the `implicit*` files in the test folder on the `dev-implicit` branch for examples on usage.
https://github.com/koka-lang/koka/tree/bf5949d1a803611be08269bd0cadf3459d22a1ec/test/type
